Post Revisions: Move UI to a modal Dialog (from Editor Sidebar)

jblz opened the pull request:
This takes the Post Revisions UI out of the Editor Sidebar & into a `Dialog` activated by the `History` button.

<img width="1327" alt="screen shot 2017-11-05 at 10 12 33 pm" src="https://user-images.githubusercontent.com/1587282/32423783-97fb6486-c276-11e7-9e36-10806c8c22be.png">

Some redux state was added and enhanced to make this possible.

## To Test

* in dev env: Open a post in the editor
* `History` button should appear as before
* Clicking `History` should open a modal dialog w/ a nice layout
* Clicking a "revision" should show the diff on the right
* <kbd>j</kbd> & <kbd>k</kbd> hotkeys should navigate through a posts revision history (h/t #19202)
* Selecting a revision & clicking `Load` should put the selected revision's title and content into the editor inputs.
* Saving after loading a revision should be sane & predictable.

### Test behind the flag

* Append the following query arg to the dev URL to load without the feature flag: `?flags=-post-editor/revisions`
* The `History` button should not appear in the Ground Control area
* There should be a revisions count in the editor's `Post Settings | Options` sidebar.
  * Clicking it should take you to the post revisions feature in wp-admin in a new tab

### Follow up work
- Improve loading placeholder ([1](https://github.com/Automattic/wp-calypso/pull/19197#discussion_r150632778))
- clean up of redundant nested sidebar 
- remove 'is_shown' class from `post-editor__inner-content` - it was added for the original implementation of revisions to allow for hiding this element and is now redundant.
- support for arbitrary revisions ([1](https://github.com/Automattic/wp-calypso/pull/19197#discussion_r150631857))
- Incorporate help info as in the wp-admin Overview, or similar ([1](https://github.com/Automattic/wp-calypso/pull/19197#issuecomment-343779418))
- Investigate and fix extrenous modal content scrollbars & fix content element size.
- Decouple `EditorRevisionsDialog` from the history button (it should be openable from anywhere) ([1](https://github.com/Automattic/wp-calypso/pull/19197#discussion_r150354371)).
- Replace `toggleDialog` with explicit open & close methods.
- Investigate tabbing/focusing between diff and revisions list - could improve keyboard workflow & accessibility
- Reconcile styling & placement of History button with the "save..." button ([1](https://github.com/Automattic/wp-calypso/pull/19197#issuecomment-342124504))
- The word "minor" alone is not great for translations or for context in English. We should consider adding the "changes" suffix back, showing empty space, or using iconography instead
- Add times to Revision list items w/o relative times so it's easier to distinguish between them

jblz commented on client/post-editor/editor-revisions-list/index.jsx:
```diff
@@ -8,86 +8,62 @@ import React, { PureComponent } from 'react';
-			isWithinBreakpoint( '>660px' )
-		) {
-			this.props.selectRevision( this.props.revisions[ 0 ].id );
+	trySelectingFirstRevision = () => {
```
this should be moved to the actions file

jblz commented on client/state/selectors/get-post-revision-changes.js:
```diff
@@ -20,7 +21,8 @@ const diffKey = ( key, obj1, obj2 ) =>
 const getPostRevisionChanges = createSelector(
 	( state, siteId, postId, revisionId ) => {
-		const orderedRevisions = getPostRevisions( state, siteId, postId, 'display' );
+		const _siteId = siteId || getSelectedSiteId( state );
```
probably take this change out since I ended up passing the `siteId` anyway

jblz commented on client/state/posts/revisions/reducer.js:
```diff
@@ -47,7 +49,22 @@ export function revisions( state = {}, action ) {
 	return state;
 }
 
+// @TODO ensure no rehydration
```
I haven't seen evidence of any. I think this is good.

jblz commented on client/state/posts/revisions/reducer.js:
```diff
@@ -47,7 +49,22 @@ export function revisions( state = {}, action ) {
+// @TODO ensure no rehydration
+export function selection( state = {}, action ) {
+	switch ( action.type ) {
+		case SELECTED_SITE_SET: {
```
alphabetize these cases

jblz commented:
@shaunandrews I added you for design review even though this isn't fully baked yet.  Please feel free to ping me if you'd like to chat about this. Suggestions (or commits) on styling greatly appreciated!

a8ci18n commented on client/post-editor/editor-revisions-list/item.jsx:
```diff
@@ -5,64 +5,71 @@
 						<span className="editor-revisions-list__additions">
-							{ this.props.translate( '%(changes)d word added', '%(changes)d words added', {
+							+
+							{ this.props.translate( '%(changes)d word', '%(changes)d words', {
```
Hi! I've found a possible matching string that has already been translated **19** times:
`translate( '%(changes)d word removed', '%(changes)d words removed' )` *ES Score: 16*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/try/post-revisions-popover)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/post-editor/editor-revisions-list/item.jsx:
```diff
@@ -5,76 +5,99 @@
-						</span>
+					{ revision.changes.added === 0 &&
+					revision.changes.removed === 0 && (
+						<span className="editor-revisions-list__minor-changes">{ translate( 'minor' ) }</span>
```
Hi! I've found a possible matching string that has already been translated **10** times:
`translate( 'Minor: ' )` *ES Score: 18*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

marekhrabe replied:
Just "minor" on its own is not very explanatory, isn't it? What about keeping "minor changes"?

a8ci18n commented on client/post-editor/editor-revisions-list/item.jsx:
```diff
@@ -5,76 +5,99 @@
 						<span className="editor-revisions-list__additions">
-							{ this.props.translate( '%(changes)d word added', '%(changes)d words added', {
+							<b>+</b>
+							{ this.props.translate( '%(changes)d', '%(changes)d', {
```
:disappointed: 19 existing translations will be lost with this change.

jblz commented on client/post-editor/post-editor.jsx:
```diff
@@ -373,8 +361,9 @@ export const PostEditor = createReactClass( {
+						// @TODO "base" revision for arbitrary comparison
 						setNestedSidebar={ this.setNestedSidebar }
-						selectRevision={ this.selectRevision }
+						selectedRevisionId={ this.state.selectedRevisionId }
```
👀

jblz commented:
> @spen RevisionsModal: Scroll to selected item in list 7217936

Nice! Looks good & works well.
thanks :)

marekhrabe commented:
I'm little overwhelmed by different colors and sizes of controls that are so close to each other. What do you think, @shaunandrews? Anything we can do? At last same font size?
    
<img width="280" alt="screen shot 2017-11-06 at 12 38 11" src="https://user-images.githubusercontent.com/156676/32439558-a4f0f59c-c2ef-11e7-8706-d0179ba5d630.png">

marekhrabe commented:
What about adding time?

<img width="231" alt="screen shot 2017-11-06 at 12 42 15" src="https://user-images.githubusercontent.com/156676/32439666-f4475078-c2ef-11e7-8d10-28aa1aaa2101.png">

marekhrabe commented:
**[BLOCKER]** Diffed titles are not styled correctly:

<img width="497" alt="screen shot 2017-11-06 at 12 49 59" src="https://user-images.githubusercontent.com/156676/32439884-0c0c6152-c2f1-11e7-8a45-218a90b2f809.png">

marekhrabe commented on client/post-editor/editor-diff-viewer/style.scss:
```diff
@@ -1,41 +1,46 @@
-	border-bottom: 2px solid $blue-wordpress;
-}
+	.editor-diff-viewer__additions {
+		background: lighten($blue-wordpress, 58%);
```
spaces inside brackets

jblz replied:
Prettier is doing this automatically. Do I need to reconfigure something?

marekhrabe replied:
if it is automatic, ignore me :) but we do prefer spaced brackets even in CSS, right?

jblz replied:
> but we do prefer spaced brackets even in CSS, right?

It [looks like we do](https://github.com/Automattic/wp-calypso/blob/27397c4b80026609c0515988bc6b73314f6a7fe6/docs/coding-guidelines/css.md#spacing).  I'll revert these lines, but it does seem like [our fork of prettier](https://github.com/Automattic/calypso-prettier) needs fixing here.

marekhrabe commented on client/post-editor/editor-diff-viewer/style.scss:
```diff
@@ -1,41 +1,46 @@
-	border-bottom: 2px solid $alert-red;
-	text-decoration: line-through;
+	.editor-diff-viewer__deletions {
+		background: lighten($alert-red, 38%);
```
spaces inside brackets

marekhrabe commented on client/post-editor/editor-revisions-list/header.jsx:
```diff
@@ -1,34 +1,37 @@
-);
+const EditorRevisionsListHeader = ( { numRevisions, translate } ) => {
+	const classes = classNames( 'editor-revisions-list__header', {
+		'editor-revisions-list__loading-placeholder': ! numRevisions,
```
As we require this prop to be a number, I think it would be cleaner to use `numRevisions === 0` and similar condition few lines under. What do you think?

jblz replied:
I'm indifferent. `! numRevisions` is more concise & `classNames` does a truthiness check anyhow.

marekhrabe commented:
Can we reset selected state on modal reopening? With revisions updating once you load them in the editor, not resetting could get you into a confusing state

a8ci18n commented on client/post-editor/editor-revisions-list/item.jsx:
```diff
@@ -5,66 +5,70 @@
-						) }
+					{ revision.changes.added === 0 &&
+					revision.changes.removed === 0 && (
+						<span className="editor-revisions-list__minor-changes">{ translate( 'minor' ) }</span>
```
Hi! I've found a possible matching string that has already been translated **10** times:
`translate( 'Minor: ' )` *ES Score: 18*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

jblz commented:
@marekhrabe @spen -- I've addressed most of the feedback (thanks again!) & this is ready for another pass.

Both of these are good points, but I'd like to address them in follow up PRs.
> I'm little overwhelmed by different colors and sizes of (the "save" & "history" buttons)...so close to each other.

> What about adding time?

Please let me know how it looks!

marekhrabe commented:
I still have few issues:
- the word "minor". I don't think it works alone and when I'm thinking about translations, I would have no way to translate this to Czech without using also word "changes" there. Can we just include it even in English?
- the styling of History button is way off
- adding also a time to the revision date (otherwise, I can be looking at tens of revisions with the exact same title)

We can address them separately to not grow this PR even more, what do you think? We are still behind the flag so it should not be a huge issue

marekhrabe commented:
There are extra scrollbars. The horizontal one shows in any browser, the vertical one in Firefox. This happens regardless of window width. It's always a nudge bigger than the modal width.

Horizontal + Vertical (Firefox):

<img width="842" alt="screen shot 2017-11-09 at 14 57 36" src="https://user-images.githubusercontent.com/156676/32609029-83f027a2-c55e-11e7-8d6e-4cb83a411b0c.png">

Just horizontal (Chrome):

![scrollbar](https://user-images.githubusercontent.com/156676/32609034-88848588-c55e-11e7-9f0a-14902b02451f.gif)

The easiest way to test on Mac: 
- go to System Preferences
- General
- Set "Show Scroll Bars" to "Always"

marekhrabe commented on client/post-editor/editor-revisions-list/item.jsx:
```diff
@@ -73,8 +77,24 @@ class EditorRevisionsListItem extends PureComponent {
-	selectRevision: PropTypes.func.isRequired,
+	siteId: PropTypes.number.isRequired,
+
+	// connected to state
```
We are trying to alphabetize all objects in our codebase but suddenly this breaks it. I've never seen this artificial categorization anywhere else in the app - is that some recent update to coding standards or completely new idea?

marekhrabe replied:
I'm not really against but it seems to break other rules we use for reviewing the code style.

spen replied:
I've started to see a lot more cases of this so started to copy... I found it useful when starting to shift towards reduxifying props since you can clearly see where they come from.
I don't feel strongly for or against though 🤷‍♂️

marekhrabe commented on client/post-editor/editor-revisions-list/style.scss:
```diff
@@ -40,50 +93,59 @@
+.editor-revisions-list__minor-changes {
+	top: 8px;
+	color: $gray-text-min;
+}
```
Minor thing but can we all update our editors to make a newline at the end of files? It helps diffs and other code manipulation. We make use of http://editorconfig.org in Calypso so it should be very easy :)

marekhrabe commented on client/post-editor/editor-revisions/dialog.jsx:
```diff
@@ -0,0 +1,118 @@
+		}
+
+		const bodyClassName = 'showing-post-revisions-dialog';
+		isVisible
```
This seems like a strange usage of the ternary operator. It's primarily designed for getting a value back, not for conditional fn calls.

marekhrabe replied:
I would turn this into an `if` statement. We cannot use `classList.toggle` as we need a support for IE 11 and they never implemented it there, only in Edge :/

spen replied:
An `if` could be a little more clear. I'm a proponent for this use of ternaries though - I feel like people are becoming much more used to ternaries than they were a year or two ago.
In terms of `classList.toggle` - I'm not a fan of that anyway. In fact I'm not a fan of `toggle` in general as I'd always prefer to explicitly add or remove. Particularly when toggle takes an argument... a true toggle should flip in every case without influence IMO.

marekhrabe commented on client/state/posts/revisions/actions.js:
```diff
@@ -1,15 +1,24 @@
+	POST_REVISIONS_SELECT,
+	POST_REVISIONS_TOGGLE_SHOWING_DIALOG,
 } from 'state/action-types';
+/*
```
Do we want to keep this for some reason or can we remove the commented section?

spen commented:
Re: @marekhrabe 

> the word "minor". I don't think it works alone and when I'm thinking about translations, I would have no way to translate this to Czech without using also word "changes" there. Can we just include it even in English?

I can't quite comment on this myself - truth be told, I don't know what 'minor' signifies. I see "+ 2" in the changes which is what I might consider 'minor'. Perhaps there's just some better way to show this altogether with an icon or something.

> the styling of History button is way off

Agreed - It's not really changed since we moved to the modal.
I'm not sure what it should look like at this point or whether or not it should even be here in the header... we can launch from anywhere.
If we are sticking with the header button it could use some improvements.

I think this is something we can hold off on and keep as follow up work.

> adding also a time to the revision date (otherwise, I can be looking at tens of revisions with the exact same title)

So `5d ago (12:36:29)` or something similar?
I don't feel strongly either way - I see the added value in granularity but it is extra noise.
I would err on the side of showing the time in some way though the extra clarity could be very useful.

Definately follow-up work

> There are extra scrollbars.

I'd say anything like this should be tackled here, they should be fairly easy to iron out and are fairly critical IMO.
I'd be happy to jump on this one.

spen commented on client/post-editor/editor-revisions-list/style.scss:
```diff
@@ -1,35 +1,88 @@
+
+	@include breakpoint( ">660px" ) {
+		border-left: 1px solid darken($sidebar-bg-color, 5%);
+		z-index: 1; // Put the list above the action-buttons:before overlay gradient. -shaun
```
Is this comment to stay? or temporary?

jblz commented on client/state/selectors/get-post-revisions.js:
```diff
@@ -14,6 +14,7 @@ import { hydrateRevision, normalizeRevision } from 'state/selectors/utils/revisi
 
 const getPostRevisions = createSelector(
 	( state, siteId, postId, normalizerName = null ) =>
+		// @TODO this should probably be gated & cached in a reducer or at least debounced
```
That it does.  Don't mind me :)

spen commented:
Is it at all possible that authors who deal exclusively (or almost exclusively) with the visual editor will find the body of these revisions confusing?

Somebody who's not versed in html, shortcodes or code in general may find some of the following a bit odd...

<img width="819" alt="screen shot 2017-11-12 at 23 41 23" src="https://user-images.githubusercontent.com/4335450/32704792-06cd9208-c803-11e7-9845-ebb56a1964f9.png">

Is it worth noting somewhere that this is the raw content? Is this a non-issue - we trust people using the editor to have some level of savvy here?

jblz commented:
> will find the body of these revisions confusing?

Undoubtedly. Once the view is working well, we should see about visual diffs (side-by-side on wide enough viewports).

jblz commented:
> Is it worth noting somewhere that this is the raw content?

I think it's fine for now...
For a follow-up: we should see about incorporating some or all of the help info provided in the wp-admin Overview section:

<img width="1335" alt="screen shot 2017-11-12 at 6 54 07 pm" src="https://user-images.githubusercontent.com/1587282/32704900-ec6e0960-c7da-11e7-8533-bb117680f490.png">

```
This screen is used for managing your content revisions.

Revisions are saved copies of your post or page, which are periodically created as you update your content. The red text on the left shows the content that was removed. The green text on the right shows the content that was added.

From this screen you can review, compare, and restore revisions:

To navigate between revisions, drag the slider handle left or right or use the Previous or Next buttons.
Compare two different revisions by selecting the “Compare any two revisions” box to the side.
To restore a revision, click Restore This Revision.

---

For more information:
[Revisions Management](https://codex.wordpress.org/Revision_Management)
[Support Forums](https://wordpress.org/support/)
```

jblz commented on client/post-editor/editor-revisions-list/item.jsx:
```diff
@@ -5,66 +5,70 @@
+							className="editor-revisions-list__deletions"
+							aria-label={ translate( '%(changes)d word removed', '%(changes)d words removed', {
+								args: { changes: revision.changes.removed },
+								count: revision.changes.removed,
```
either way -- `get` seems more straightforward to me. Feel free to add it

jblz commented on client/post-editor/editor-ground-control/history-button.jsx:
```diff
@@ -11,79 +11,40 @@ import { flow } from 'lodash';
+		>
+			{ translate( 'History' ) }
+		</button>
+		<EditorRevisionsDialog
```
Ya, think it can be recomposed higher up the call stack.  Let's note it for a follow up since it may have side effects at this point.

jblz commented on client/post-editor/editor-ground-control/history-button.jsx:
```diff
@@ -11,79 +11,40 @@ import { flow } from 'lodash';
+	<div className="editor-ground-control__history">
+		<button
+			className="editor-ground-control__history-button button is-link"
+			onClick={ toggleDialog }
```
+1.

spen commented on client/post-editor/editor-revisions/style.scss:
```diff
@@ -1,17 +1,25 @@
+}
+
+.editor-revisions__wrapper {
+	height: 100vh;
```
We might want to revisit these dims (`100vh`/`100vw`) as they actually cause the modal area to extend off page (not visually though):

<img width="1783" alt="screen shot 2017-11-13 at 02 44 25" src="https://user-images.githubusercontent.com/4335450/32707565-b53fe994-c81c-11e7-8cb6-acf043be2c69.png">

marekhrabe replied:
Good catch. Let's switch to % and make sure it computes against the modal size (`position: relative` on modal)

jblz replied:
I'll see if I can solve this quickly.  If not, going to move to a follow up PR

jblz replied:
This is kind of fidgety given the interaction w/ the Dialog styles.  Let's fix this in a follow up PR.

a8ci18n commented on client/post-editor/editor-revisions-list/item.jsx:
```diff
@@ -5,65 +5,72 @@
-							</span>
+					{ added === 0 &&
+						removed === 0 && (
+							<span className="editor-revisions-list__minor-changes">{ translate( 'minor' ) }</span>
```
Hi! I've found a possible matching string that has already been translated **10** times:
`translate( 'Minor: ' )` *ES Score: 18*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n replied:
Hi! I've found a possible matching string that has already been translated **10** times:
`translate( 'Minor: ' )` *ES Score: 18*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n replied:
Hi! I've found a possible matching string that has already been translated **10** times:
`translate( 'Minor: ' )` *ES Score: 18*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

yoavf replied:
@jblz this string is a bit ambiguous  - can you add a translator comment specifying that it's about revisions?

spen commented:
🎉

yoavf commented on client/post-editor/editor-revisions/dialog.jsx:
```diff
@@ -0,0 +1,121 @@
+				compact: true,
+				disabled: ! ( revision && postId && siteId ),
+				isPrimary: true,
+				label: translate( 'Load' ),
```
Same here  - translators would need to know what is being loaded to properly translate this.
